### PR TITLE
ci: automate package version updates

### PR DESCRIPTION
**Pull request by quinna-h**
- Calls `scripts/freshvenvs` to generate a list of "outdated" packages (packages where we don't test the full range from MIN_SUPPORTED_VERSION to MAX_SUPPORTED_VERSION)
- Automate regenerating riot requirements.txt files for each out-of-date packages
- Gets called in GH workflow weekly cron job
         - Can adjust frequency / time - currently Sunday midnight
- TODO: open a new PR for every venv with outdated versions - right now it stops at one. 

Testing:

Github Actions workflows: https://github.com/DataDog/dd-trace-py/actions/workflows/generate-package-versions.yml
Example PR: https://github.com/DataDog/dd-trace-py/pull/10333

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: .github/workflows/generate-package-versions.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.10"
+  
+    - name: Run regenerate-riot-latest
+      run: ./scripts/regenerate-riot-latest.sh
+
+    - name: Create Pull Request
+      id: pr
+      uses: peter-evans/create-pull-request@v6
```
Comment (quinna-h) [inline .github/workflows/generate-package-versions.yml:27]: Documentation here: https://github.com/peter-evans/create-pull-request?tab=readme-ov-file#documentation

File: scripts/update-latest.py
[the remark this answers is not in the document]
Comment (quinna-h) [reply scripts/update-latest.py:1]: Thank you for the suggestion! I've changed the script to use this

File: scripts/regenerate-riot-latest.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        BRANCH_NAME="upgrade-${pkg}-$(date +%Y%m%d%H%M%S)"
+        git checkout -b "$BRANCH_NAME"
+        for h in ${RIOT_HASHES[@]}; do 
+            rm ".riot/requirements/${h}.txt"
+        done
+        scripts/compile-and-prune-test-requirements
+
+        git add .
+        git commit -m "Upgrade version for $pkg"
+        git push origin "$BRANCH_NAME"
```
Comment (bouwkast) [inline scripts/regenerate-riot-latest.sh:31]: I don't think we want to do `git` commands here

File: scripts/regenerate-riot-latest.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+pkgs=$(python scripts/freshvenvs.py | cut -d':' -f1)
+echo $pkgs
+
+if ! $DDTEST_CMD; then
+    echo "Command '$DDTEST_CMD' failed."
+    exit 1
+fi
+
+for pkg in ${pkgs[*]}; do
```
Comment (bouwkast) [inline scripts/regenerate-riot-latest.sh:14]: What if we limit this to just a few packages per PR and then have another manual step that can do the entire list of packages?

File: scripts/freshvenvs.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+import sys
+import typing
+
+from packaging.version import Version
+from pip import _internal
+
+
+sys.path.append(str(pathlib.Path(__file__).parent.parent.resolve()))
+
+CONTRIB_ROOT = "ddtrace/contrib"
```
Comment (P403n1x87) [inline scripts/freshvenvs.py:18]:
```suggestion
CONTRIB_ROOT = pathlib.Path("ddtrace/contrib")
```

File: scripts/freshvenvs.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        self.extend(self._stringio.getvalue().splitlines())
+        del self._stringio  # free up some memory
+        sys.stdout = self._stdout
+        sys.stderr = self._stderr
+
+
+def _get_integrated_modules() -> typing.Set[str]:
+    """Get all modules that have contribs implemented for them"""
+    all_required_modules = set()
+    for item in os.listdir(CONTRIB_ROOT):
```
Comment (P403n1x87) [inline scripts/freshvenvs.py:39]:
```suggestion
    for item in CONTRIB_ROOT.iterdir():
```

Comment (P403n1x87) [reply scripts/freshvenvs.py:39]: Then `item` becomes `contrib_dir`

File: scripts/freshvenvs.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        contrib_dir = f"{CONTRIB_ROOT}/{item}"
+        init_filepath = f"{contrib_dir}/__init__.py"
```
Comment (P403n1x87) [inline scripts/freshvenvs.py:41]:
```suggestion
        init_filepath = item / "__init__.py"
```

File: scripts/freshvenvs.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        if version_age > dt.timedelta(days=365 * 2):
+            earliest_within_window = version
+            break
+    return earliest_within_window, versions[0]
+
+
+def _get_package_versions_from(env: str, packages: typing.Set[str]) -> typing.List[typing.Tuple[str, str]]:
+    """Return the list of package versions that are tested"""
+    with open(f".riot/requirements/{env}.txt", "r") as lockfile:
+        lockfile_content = lockfile.readlines()
```
Comment (P403n1x87) [inline scripts/freshvenvs.py:107]:
```suggestion
    lockfile_content = pathlib.Path(f".riot/requirements/{env}.txt").read_text().splitlines()
```

File: scripts/freshvenvs.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    lock_packages = []
+    for line in lockfile_content:
+        parts = line.split("==")
```
Comment (P403n1x87) [inline scripts/freshvenvs.py:110]:
```suggestion
        package, _, version = line.partition("==")
```

**Review by P403n1x87 — COMMENTED** [review]
just a few nits

File: scripts/regenerate-riot-latest.sh [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    echo "Checking if new latest version exists for $pkg"
+    export VENV_NAME="$pkg"
+    RIOT_HASHES=( $(riot list --hash-only "^${VENV_NAME}$") )
```
Comment (emmettbutler) [inline scripts/regenerate-riot-latest.sh:17]: This seems to rely on the assumption that the name of an integrated package will always also be the name of the relevant riot environment, which is not always the case. https://github.com/DataDog/dd-trace-py/blob/e94ed6049e2addab48d51fbbc1404f3995a3c941/ddtrace/_monkey.py#L115

**Review by emmettbutler — APPROVED** [review]
This is great progress. I mentioned a bit of complexity that will need to be addressed, but I don't think it needs to happen before this change is merged.

File: .github/workflows/generate-package-versions.yml [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+      - name: Create Pull Request
+        id: pr
+        uses: peter-evans/create-pull-request@v6
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          branch: "upgrade-${{ env.VENV_NAME }}-version"
+          commit-message: "Test Package Versions update"
+          delete-branch: true
+          base: main
+          title: "chore: update ${{ env.VENV_NAME }} latest version"
```
Comment (bouwkast) [inline .github/workflows/generate-package-versions.yml:81]:
```suggestion
          title: "chore: update ${{ env.VENV_NAME }} to the latest version"
```

nitpicky - feel free to ignore

**Review by bouwkast — APPROVED** [review]
Great work!

(Defer to the python experts for python code review :smile:)

I just had a few nitpicks on the generated PR title/body - real small stuff

**Pull request merged**